v1/ast/term.go

```diff
@@ -2697,31 +2691,28 @@ func (obj *object) insert(k, v *Term, resetSortGuard bool) {
 
 	for curr := head; curr != nil; curr = curr.next {
 		if equal(curr.key.Value) {
-			// The ground bit of the value may change in
-			// replace, hence adjust the counter per old
-			// and new value.
-
 			if curr.value.IsGround() {
 				obj.ground--
 			}
 			if v.IsGround() {
 				obj.ground++
 			}
 
+			// Update hash based on the new value
 			curr.value = v
+			obj.elems[hash] = curr
+			obj.hash = 0
+			for hash := range obj.elems {
+				obj.hash += hash + obj.elems[hash].value.Hash()
+			}
 
-			obj.rehash()
 			return
 		}
 	}
-	elem := &objectElem{
-		key:   k,
-		value: v,
-		next:  head,
-	}
-	obj.elems[hash] = elem
+
+	obj.elems[hash] = &objectElem{key: k, value: v, next: head}
 	// O(1) insertion, but we'll have to re-sort the keys later.
-	obj.keys = append(obj.keys, elem)
+	obj.keys = append(obj.keys, obj.elems[hash])
```

Comment:
The extra lookup is cheaper than GC:ing the local pointer var? Also for large objects?

Reply:
Same question for the loop above.